gnome-initial-setup/gis-driver.c

```diff
@@ -288,6 +294,118 @@ gis_driver_get_mode (GisDriver *driver)
 }
 
 gboolean
+gis_driver_is_in_demo_mode (GisDriver *driver)
+{
+  GisDriverPrivate *priv = gis_driver_get_instance_private (driver);
+  return priv->is_in_demo_mode;
+}
+
+#define DEMO_ACCOUNT_USERNAME "demo-guest"
+#define DEMO_ACCOUNT_FULLNAME "Demo"
+
+static void
+handle_demo_mode_error (GError *error)
+{
+  GtkWidget *dialog;
+
+  g_warning ("%s", error->message);
+  dialog = gtk_message_dialog_new (NULL,
+                                   GTK_DIALOG_DESTROY_WITH_PARENT,
+                                   GTK_MESSAGE_ERROR,
+                                   GTK_BUTTONS_CLOSE,
+                                   "Failed to enter demo mode: %s",
+                                   error->message);
+  gtk_dialog_run (GTK_DIALOG (dialog));
+  gtk_widget_destroy (dialog);
+  g_error_free (error);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (ActUser, g_object_unref)
+
+static gboolean
+create_demo_user (GisDriver *driver)
+{
+  GisDriverPrivate *priv = gis_driver_get_instance_private (driver);
+  g_autoptr(ActUser) user;
+  GError *error = NULL;
+  const gchar *language;
+
+  error = NULL;
+  user = act_user_manager_create_user (act_user_manager_get_default (),
+                                       DEMO_ACCOUNT_USERNAME,
+                                       DEMO_ACCOUNT_FULLNAME,
+                                       ACT_USER_ACCOUNT_TYPE_ADMINISTRATOR,
+                                       &error);
+  if (error)
+    {
+      handle_demo_mode_error (error);
+      return FALSE;
+    }
+
+  act_user_set_password_mode (user, ACT_USER_PASSWORD_MODE_NONE);
+  act_user_set_automatic_login (user, FALSE);
```

Comment:
Ah of course! Something to note here as well is that "automatic login" only works for the first time the greeter is shown after GDM starts up https://github.com/endlessm/gdm/blob/a291d18083c7a420f701aef896ced58ad95fcc1e/daemon/gdm-manager.c#L1263 , so the typical automatic login mode may not be what we want. Timed login tends to fit this purpose better since it runs every time. If AccountsService exposes a mechanism to set up timed login then I'd be glad to use that and drop my other pkexec script.